feat(backend): add search on /book/

agory opened the pull request:
➕ add findByAuthorAndTitle in BookRepository
➕ update BookFacade and BookController to handle New RequestParam
➕ update test and fix Mockito.times(1)

bricebuffa commented on src/main/java/com/example/bookstore/dto/QueryBookVO.java:
```diff
@@ -0,0 +1,23 @@
+        this.title = title;
+    }
+
+    public boolean isEmpty() {
```
nom ambigu on pourrait penser que c'est le getter d'un attribut booléen

agory replied:
done

bricebuffa commented on src/main/java/com/example/bookstore/facades/BookFacade.java:
```diff
@@ -19,8 +20,16 @@
 
-    public List<BookShortVO> retrieveBookList() {
-        List<Book> books = this.bookRepository.findAll();
+    public List<BookShortVO> retrieveBookList(QueryBookVO queryBookVO) {
```
je pense qu'il faut de la doc ici car tu ne récupères plus seulement la liste des livres telle quelle mais tu les filtres

agory replied:
done

bricebuffa replied:
not done ?

agory replied:
dsl j'ai confondu avec la methode du controller. Je met à jour ça tout de suite

agory replied:
done

bricebuffa commented on src/test/java/com/example/bookstore/facades/BookFacadeTest.java:
```diff
@@ -27,40 +28,72 @@
-    @Test
-    public void retrieveBookList_shouldReturnAListOfBookShort() {
-        // given
+    private List<Book> getBookList() {
```
Fais une méthode setup() {...} précédé de @Before dans laquelle tu initialises une liste de livre et un livre qui sont des attributs privée de ta classe de teste pour remplacer getBookLlist et getBook

bricebuffa commented on src/test/java/com/example/bookstore/facades/BookFacadeTest.java:
```diff
@@ -27,40 +28,72 @@
+    @Test
+    public void retrieveBookList_shouldReturnAListOfBookShort_givenEmptyQueryBook() {
+        // given
+        QueryBookVO queryBookVO = new QueryBookVO(null,null);
```
utilise un mock => mock(QueryBookVO.class);

agory replied:
done

agory committed:
feat(backend): add search on /book/

Fix some test setup @before.
 Add dto QueryBook.
 update book controller and facade to handle
new RequestParam author and title.
 Add a new query in Book repository to find Author and title in
book

bricebuffa commented on src/main/java/com/example/bookstore/repositories/BookRepository.java:
```diff
@@ -14,4 +16,13 @@
+        "where upper(b.authors) like concat('%', upper(:author), '%') " +
+        "and upper(b.title) like concat('%', upper(:title), '%') " +
+        "order by b.publishDate desc ")
+    List<Book> findByTitleAndAuthor(
```
Doc

agory replied:
done

bricebuffa commented on src/main/java/com/example/bookstore/facades/BookFacade.java:
```diff
@@ -19,8 +20,21 @@
-        List<Book> books = this.bookRepository.findAllByOrderByPublishDateDesc();
+    /***
+     * Retrieve all books or retrieve all book that match title et author in queryBookVO
+     * @param queryBookVO
```
les params ne sont pas nécessaire si il y n'y a pas de description, donc si pas de besion de description supplémentaire ne renseigne pas les @params et @return car ils sont strictement redondant avec la signature de la méthode
idem plus bas
